kernel/comps/systree/src/writer.rs

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::fmt::{Arguments, Write};
+
+use ostd::mm::{FallibleVmWrite, VmReader, VmWriter};
+
+use crate::{Error, Result};
+
+/// A specialized writer for kernel-generated content.
+///
+/// `ContextWriter` is designed to handle the common pattern in kernel where kernel
+/// code needs to generate formatted text output (like status information, statistics,
+/// or configuration data) and write it to user space with proper offset handling.
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// fn read_attr(&self, name: &str, mut writer: ContextWriter) -> Result<usize> {
+///     match name {
+///         "status" => {
+///             writeln!(writer, "Status: {}", self.status())?;
+///             writeln!(writer, "Uptime: {} seconds", self.uptime())?;
+///         }
+///         "config" => {
+///             writeln!(writer, "max_connections: {}", self.max_connections)?;
+///             writeln!(writer, "timeout: {}", self.timeout)?;
+///         }
+///         _ => return Err(Error::NotFound),
+///     }
+///
+///     Ok(writer.bytes_written())
+/// }
+/// ```
+pub struct ContextWriter<'a, 'b> {
+    /// The underlying [`VmWriter`] to write the final output.
+    writer: &'a mut VmWriter<'b>,
+    /// Number of bytes to skip from the beginning of the output.
+    ///
+    /// When content is written through this writer, the first `bytes_to_skip`
+    /// bytes will be discarded, and only subsequent bytes will be written
+    /// to the underlying `VmWriter`.
+    bytes_to_skip: usize,
+    /// Total number of bytes written to the underlying writer.
+    bytes_written: usize,
+}
+
+impl<'a, 'b> ContextWriter<'a, 'b> {
+    /// Creates a new `ContextWriter` with the specified underlying writer and skip offset.
+    pub fn new(writer: &'a mut VmWriter<'b>, bytes_to_skip: usize) -> Self {
+        Self {
+            bytes_to_skip,
+            writer,
+            bytes_written: 0,
+        }
+    }
+
+    /// Returns the total number of bytes written to the underlying writer.
+    pub fn bytes_written(&self) -> usize {
+        self.bytes_written
+    }
+
+    /// Writes formatted content to the underlying writer.
+    pub fn write_fmt(&mut self, args: Arguments<'_>) -> Result<usize> {
+        let bytes_before = self.bytes_written;
+        Write::write_fmt(self, args).map_err(|_| Error::AttributeError)?;
```

Comment:
If fmt::Write is in scope, will `writer.write_fmt` call the method associated to the type? Or is it ambiguous and lead to compiler errors?